### PR TITLE
[SPARK-50166][SQL][BUILD] Fix shade and relocation rule of `sql/core` module

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr makes the following changes to the `maven-shade-plugin` rules for the `sql/core` module:

1. To avoid being influenced by the parent `pom.xml`, use `combine.self = "override"` in the `<configuration>` of the `maven-shade-plugin` for the `sql/core` module. Before this configuration was added, the relocation result was incorrect, and `protobuf-java` was not relocated. We can unzip the packaging result to confirm this issue.

We can use IntelliJ's "Show Effective POM" feature to view the result of this parameter, the result is equivalent to the effective POM log with --debug printing added during the Maven compilation:

**Before**

image

We can see that an unexpected 
```
<includes>
  <include>org.eclipse.jetty.**</include>
</includes>
```
 has been added to the relocation rule.

**After** 

image

We can see that the extra `<includes>` in the relocation rule is no longer present.

2. Before SPARK-48755 | https://github.com/apache/spark/pull/47133 overwrote the `maven-shade-plugin` rules for `sql/core`, it inherited the rules from the parent `pom.xml` and shaded `org.spark-project.spark:unused`. This behavior changed after SPARK-48755, so this pr restores it.

3. The relocation rules for Guava should be retained and follow the configuration in the parent `pom.xml`, which relocates `com.google.common` to `${spark.shade.packageName}.guava`. This PR restores this configuration.

4. For `protobuf-java`, which is under the `com.google.protobuf` package, the already shaded `protobuf-java` in the `core` module can be reused instead of shading it again in `sql/core` module. Therefore, this pr only configures the corresponding relocation rule for it: `com.google.protobuf` -> `${spark.shade.packageName}.spark_core.protobuf`.

5. Regarding the `ServicesResourceTransformer` configuration, it is used to merge `META-INF/services` resources. This is not needed for Guava and `protobuf-java`, so this pr removes it.

### Why are the changes needed?
Fix shade and relocation rule of `sql/core` module

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Pass Github Aciton
- Manually inspect the packaging result: Extract `spark-sql_2.13-4.0.0-SNAPSHOT.jar` to a separate directory, then execute `grep "org.sparkproject.guava" -R *` and `grep "org.sparkproject.spark_core.protobuf" -R *` to confirm the successful relocation.
- Maven test passed: https://github.com/LuciferYang/spark/runs/32278520082

image

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/core/pom.xml [hunk 1]
```diff
@@ -277,22 +277,22 @@
...
           </artifactSet>
           <relocations>
             <relocation>
-              <pattern>com.google</pattern>
-              <shadedPattern>${spark.shade.packageName}.com.google</shadedPattern>
-              <excludes>
-                <!-- Guava is relocated to ${spark.shade.packageName}.connect.guava -->
-                <exclude>com.google.common.**</exclude>
-              </excludes>
+              <pattern>com.google.common</pattern>
```
Comment (LuciferYang) [inline sql/core/pom.xml:285]: cc @pan3793

Comment (pan3793) [reply sql/core/pom.xml:285]: the effective relocation rules are very confusing, seems to be affected by parent pom

Comment (LuciferYang) [reply sql/core/pom.xml:285]: We can use `combine.self = "override"` to resolve this issue.

**Review by pan3793 — APPROVED** [review]
Thanks for fixing this.

**Comment by LuciferYang** [comment]
also cc @HyukjinKwon @dongjoon-hyun @HeartSaVioR @bogao007 FYI

**Comment by LuciferYang** [comment]
Thanks @pan3793 ~

**Review by dongjoon-hyun — APPROVED** [review]
According to the PR description, this is a regression due to SPARK-48755 ?
- https://github.com/apache/spark/pull/47133

Thank you for discovering this regression and fixing before Apache Spark 4.0.0 release, @LuciferYang .

**Comment by dongjoon-hyun** [comment]
BTW, to @LuciferYang , we need to wait for [SPARK-48755](https://issues.apache.org/jira/browse/SPARK-48755) people as you pinged already.

**Comment by LuciferYang** [comment]
> According to the PR description, this is a regression due to [SPARK-48755](https://issues.apache.org/jira/browse/SPARK-48755) ?
> 
> * [[SPARK-48755][SS][PYTHON] transformWithState pyspark base implementation and ValueState support #47133](https://github.com/apache/spark/pull/47133)
> 
> Thank you for discovering this regression and fixing before Apache Spark 4.0.0 release, @LuciferYang .

Yes

**Comment by LuciferYang** [comment]
> BTW, to @LuciferYang , we need to wait for [SPARK-48755](https://issues.apache.org/jira/browse/SPARK-48755) people as you pinged already.

Got it ~

**Comment by dongjoon-hyun** [comment]
Gentle ping once more, @bogao007 and @HeartSaVioR , because the change is reported at https://github.com/apache/spark/pull/47133 .

**Comment by bogao007** [comment]
I don't have specific concerns as long as the existing CI passes, thanks!

**Review by HeartSaVioR — APPROVED** [review]
+1 thanks for the fix!

**Comment by LuciferYang** [comment]
Merged into master for Spark 4.0. Thanks @pan3793 @dongjoon-hyun @HeartSaVioR and @bogao007 ~

**Pull request closed without merging**